### PR TITLE
Improve current generator implementation

### PR DESCRIPTION
**Pull request by djigitiniot**
https://trendz-tb.atlassian.net/browse/TB-1563

### CODE REVIEW

**Comment by Deesthortered** [comment]
Use logback.xml for log management, not application.yml

**Comment by djigitiniot** [comment]
> Use logback.xml for log management, not application.yml

done

**Comment by Deesthortered** [comment]
1. Better to not use "var" where it is not really needed, better not to use than to use.
    Example when it is needed: var reference = new ParameterizedTypeReference...
    Examples where it is not needed:
       * var objectMapper = ...
       * var customerOpt = ...
       * var response = execution.execute(...

2. A lot of not necessary "final", use effective final effect.
3. If you have a function with a lot of arguments, better to make a new row, do not do column.
    Example with column (worse):
    private ModelData makeData(boolean skipTelemetry,
                                                  ZonedDateTime startYear,
                                                  boolean fullTelemetryGeneration,
                                                  long startGenerationTime,
                                                  long endGenerationTime, ....) { ... }

    Example with row (better):
    private ModelData makeData(
           boolean skipTelemetry, ZonedDateTime startYear, boolean fullTelemetryGeneration, 
           long startGenerationTime, long endGenerationTime, .....
    ) { ... }

File: src/main/java/org/thingsboard/trendz/generator/SolutionTemplateGeneratorApplication.java [hunk 1]
```diff
@@ -1,14 +1,18 @@
...
 import org.springframework.boot.autoconfigure.SpringBootApplication;
+import org.springframework.context.ApplicationContext;
 import org.thingsboard.trendz.generator.exception.SolutionValidationException;
 import org.thingsboard.trendz.generator.solution.SolutionTemplateGenerator;
+import org.thingsboard.trendz.generator.utils.DateTimeUtils;
 import org.thingsboard.trendz.generator.utils.RandomUtils;
 
+import java.time.ZoneId;
 import java.time.ZonedDateTime;
 import java.time.format.DateTimeFormatter;
```
Comment (Deesthortered) [inline src/main/java/org/thingsboard/trendz/generator/SolutionTemplateGeneratorApplication.java]: remove unused import

File: src/main/resources/application.yml [hunk 2]
```diff
@@ -1,5 +1,6 @@
-server:
-  port: 8887
+spring:
+  main:
+    web-application-type: NONE
```
Comment (Deesthortered) [inline src/main/resources/application.yml:3]: nice!

File: src/main/java/org/thingsboard/trendz/generator/solution/watermetering/WaterMeteringSolution.java [hunk 3]
```diff
@@ -762,17 +846,37 @@ private PumpStation makePumpStationByRegion(PumpStationConfiguration pumpStation
...
             return new Telemetry<>("skip");
         }
 
         Telemetry<Long> result = new Telemetry<>("consumption");
         ConsumerType type = consumerConfiguration.getType();
 
-        long now = System.currentTimeMillis();
-        ZonedDateTime startDate = startYear.truncatedTo(ChronoUnit.HOURS);
-        ZonedDateTime nowDate = DateTimeUtils.fromTs(now).truncatedTo(ChronoUnit.HOURS);
+        long fromMs = startYear.toInstant().toEpochMilli();
```
Comment (Deesthortered) [inline src/main/java/org/thingsboard/trendz/generator/solution/watermetering/WaterMeteringSolution.java]: Use DateTimeUtils.fromTs()

File: src/main/java/org/thingsboard/trendz/generator/solution/watermetering/WaterMeteringSolution.java [hunk 4]
```diff
@@ -762,17 +846,37 @@ private PumpStation makePumpStationByRegion(PumpStationConfiguration pumpStation
...
+        long toMs = ZonedDateTime.now(ZoneId.of("UTC")).toInstant().toEpochMilli();
```
Comment (Deesthortered) [inline src/main/java/org/thingsboard/trendz/generator/solution/watermetering/WaterMeteringSolution.java]: just use System.currentTimeMillis() - it is already UTC

File: src/main/java/org/thingsboard/trendz/generator/utils/DateTimeUtils.java [hunk 5]
```diff
@@ -1,6 +1,7 @@
 package org.thingsboard.trendz.generator.utils;
 
 import org.thingsboard.trendz.generator.model.tb.Timestamp;
+import oshi.util.tuples.Pair;
```
Comment (Deesthortered) [inline src/main/java/org/thingsboard/trendz/generator/utils/DateTimeUtils.java]: use import org.apache.commons.lang3.tuple.Pair;

**Review by Deesthortered — CHANGES_REQUESTED** [review]

**Comment by Deesthortered** [comment]
The default Zone is UTC, so don't specify ZoneId like that: DateTimeUtils.fromTs(fromMs, ZoneId.of("UTC")), just use DateTimeUtils.fromTs(fromMs)

**Comment by Deesthortered** [comment]
All usages of DateTimeUtils.getDatesIntersection extract in separate method that will return pair of ts (from, to)

File: src/main/java/org/thingsboard/trendz/generator/SolutionTemplateGeneratorApplication.java (hunk 1 above)
Comment (djigitiniot) [reply src/main/java/org/thingsboard/trendz/generator/SolutionTemplateGeneratorApplication.java]: done

File: src/main/java/org/thingsboard/trendz/generator/SolutionTemplateGeneratorApplication.java [hunk 6]
```diff
@@ -1,85 +1,103 @@
...
 
 @Slf4j
 @SpringBootApplication
 public class SolutionTemplateGeneratorApplication implements CommandLineRunner {
 
-	private final static String MODE_GENERATE = "generate";
-	private final static String MODE_REMOVE = "remove";
-	private final static DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss z");
+    @Autowired
+    private ApplicationContext cnt;
```
Comment (Deesthortered) [inline src/main/java/org/thingsboard/trendz/generator/SolutionTemplateGeneratorApplication.java]: remove if not used

File: src/main/java/org/thingsboard/trendz/generator/solution/SolutionTemplateGenerator.java [hunk 7]
```diff
@@ -10,7 +10,8 @@ public interface SolutionTemplateGenerator {
 
     void validate() throws SolutionValidationException;
 
-    void generate(boolean skipTelemetry, ZonedDateTime startYear);
+    void generate(boolean skipTelemetry, ZonedDateTime startYear, boolean strictGeneration, boolean fullTelemetryGeneration,
```
Comment (Deesthortered) [inline src/main/java/org/thingsboard/trendz/generator/solution/SolutionTemplateGenerator.java]: not like this

File: src/main/java/org/thingsboard/trendz/generator/solution/watermetering/WaterMeteringSolution.java [hunk 8]
```diff
@@ -1136,4 +1206,21 @@ private String getConsumerConsumptionFileName(ConsumerType type) {
                 throw new IllegalArgumentException("Unsupported type: " + type);
         }
     }
+
+    private Pair<Long, Long> calculateNewDateRange(long from, long to, long startGenerationTime, long endGenerationTime, boolean fullTelemetryGeneration)
```
Comment (Deesthortered) [inline src/main/java/org/thingsboard/trendz/generator/solution/watermetering/WaterMeteringSolution.java]: make row for arguments

File: src/main/java/org/thingsboard/trendz/generator/solution/watermetering/WaterMeteringSolution.java [hunk 9]
```diff
@@ -749,10 +807,13 @@ private Consumer makeConsumerByConfiguration(RegionConfiguration regionConfigura
                 .build();
     }
 
-    private PumpStation makePumpStationByRegion(PumpStationConfiguration pumpStationConfiguration, Region region) {
+    private PumpStation makePumpStationByRegion(PumpStationConfiguration pumpStationConfiguration, Region region, boolean fullTelemetryGeneration) {
```
Comment (Deesthortered) [inline src/main/java/org/thingsboard/trendz/generator/solution/watermetering/WaterMeteringSolution.java:817]: same

File: src/main/java/org/thingsboard/trendz/generator/solution/watermetering/WaterMeteringSolution.java [hunk 10]
```diff
@@ -731,15 +785,19 @@ private Region makeRegionByConfiguration(RegionConfiguration regionConfiguration
                 .build();
     }
 
-    private Consumer makeConsumerByConfiguration(RegionConfiguration regionConfiguration, ConsumerConfiguration consumerConfiguration, boolean skipTelemetry) {
+    private Consumer makeConsumerByConfiguration(RegionConfiguration regionConfiguration, ConsumerConfiguration consumerConfiguration, boolean skipTelemetry, boolean fullTelemetryGeneration, long startGenerationTime, long endGenerationTime) {
```
Comment (Deesthortered) [inline src/main/java/org/thingsboard/trendz/generator/solution/watermetering/WaterMeteringSolution.java:795]: same

File: src/main/java/org/thingsboard/trendz/generator/solution/watermetering/WaterMeteringSolution.java [hunk 11]
```diff
@@ -709,10 +759,11 @@ private City makeCityByConfiguration(CityConfiguration cityConfiguration, boolea
                 .build();
     }
 
-    private Region makeRegionByConfiguration(RegionConfiguration regionConfiguration, boolean skipTelemetry) {
+    private Region makeRegionByConfiguration(RegionConfiguration regionConfiguration, boolean skipTelemetry, boolean fullTelemetryGeneration, long startGenerationTime, long endGenerationTime) {
```
Comment (Deesthortered) [inline src/main/java/org/thingsboard/trendz/generator/solution/watermetering/WaterMeteringSolution.java:769]: same

File: src/main/java/org/thingsboard/trendz/generator/solution/greenhouse/GreenhouseSolution.java [hunk 12]
```diff
@@ -2689,8 +2875,18 @@ private void createTelemetryConsumptionEnergy(Telemetry<Integer> insideLightTele
                 .flatMap(Collection::stream)
                 .collect(Collectors.groupingBy(Function.identity(), Collectors.counting()));
 
-        ZonedDateTime startDate = DateTimeUtils.fromTs(configuration.getStartTs()).truncatedTo(ChronoUnit.HOURS);
-        ZonedDateTime endDate = DateTimeUtils.fromTs(configuration.getEndTs()).truncatedTo(ChronoUnit.HOURS);
+        long fromMs = configuration.getStartTs();
```
Comment (Deesthortered) [inline src/main/java/org/thingsboard/trendz/generator/solution/greenhouse/GreenhouseSolution.java]: unused

File: src/main/java/org/thingsboard/trendz/generator/utils/DateTimeUtils.java [hunk 13]
```diff
@@ -27,4 +28,15 @@ public static ZonedDateTime fromTs(Timestamp ts, ZoneId zoneId) {
     public static long toTs(ZonedDateTime zonedDateTime) {
         return zonedDateTime.toInstant().toEpochMilli();
     }
+
+    public static Pair<Long, Long> getDatesIntersection(long startDate, long endDate, long startGenerationDate, long endGenerationDate) {
+        final long latestStartDate = Math.max(startDate, startGenerationDate);
+        final long earliestEndDate = Math.min(endDate, endGenerationDate);
+
+        if (earliestEndDate < latestStartDate) {
```
Comment (Deesthortered) [inline src/main/java/org/thingsboard/trendz/generator/utils/DateTimeUtils.java:39]: what if equal

File: src/main/java/org/thingsboard/trendz/generator/solution/energymetering/EnergyMeteringSolution.java [hunk 14]
```diff
@@ -770,7 +801,9 @@ private Building makeBuildingByConfiguration(BuildingConfiguration configuration
                         .computeIfAbsent(floor, key -> new HashMap<>())
                         .computeIfAbsent(number, key -> configuration.getDefaultApartmentConfiguration());
 
-                Apartment apartment = createApartmentByConfiguration(apartmentConfiguration, configuration.getName(), floor, number, configuration.getApartmentsByFloorCount(), skipTelemetry);
+                Apartment apartment = createApartmentByConfiguration(
+                        apartmentConfiguration, configuration.getName(), floor, number, configuration.getApartmentsByFloorCount(), skipTelemetry,
+                        fullTelemetryGeneration, startGenerationTime, endGenerationTime);
```
Comment (Deesthortered) [inline src/main/java/org/thingsboard/trendz/generator/solution/energymetering/EnergyMeteringSolution.java:815]: Almost - ")" always on the new row, add \n after "number" arg

File: src/main/java/org/thingsboard/trendz/generator/solution/energymetering/EnergyMeteringSolution.java [hunk 15]
```diff
@@ -630,38 +656,43 @@ private ModelData makeData(boolean skipTelemetry, ZonedDateTime startYear) {
         );
 
         Set<ModelEntity> buildings = buildingConfigurations.stream()
-                .map(configuration -> makeBuildingByConfiguration(configuration, skipTelemetry))
+                .map(configuration -> makeBuildingByConfiguration(
+                        configuration, skipTelemetry, fullTelemetryGeneration, startGenerationTime, endGenerationTime))
```
Comment (Deesthortered) [inline src/main/java/org/thingsboard/trendz/generator/solution/energymetering/EnergyMeteringSolution.java:666]: same

**Review by Deesthortered — CHANGES_REQUESTED** [review]

**Comment by Deesthortered** [comment]
deleteRuleChain() -> findAny() - why any? why not all?

File: src/main/java/org/thingsboard/trendz/generator/utils/DateTimeUtils.java (hunk 13 above)
Comment (djigitiniot) [reply src/main/java/org/thingsboard/trendz/generator/utils/DateTimeUtils.java:39]: if both are equal, then both are returned in Pair

File: src/main/java/org/thingsboard/trendz/generator/solution/energymetering/EnergyMeteringSolution.java (hunk 14 above)
Comment (djigitiniot) [reply src/main/java/org/thingsboard/trendz/generator/solution/energymetering/EnergyMeteringSolution.java:815]: how many arguments do you want to be in separate row ?

Comment (djigitiniot) [reply src/main/java/org/thingsboard/trendz/generator/solution/energymetering/EnergyMeteringSolution.java:815]: > Almost - ")" always on the new row

what ?

File: src/main/java/org/thingsboard/trendz/generator/solution/SolutionTemplateGenerator.java (hunk 7 above)
Comment (djigitiniot) [reply src/main/java/org/thingsboard/trendz/generator/solution/SolutionTemplateGenerator.java]: like how ?

**Comment by djigitiniot** [comment]
> deleteRuleChain() -> findAny() - why any? why not all?

you went the same logic in BasicSolution:remove method

```
tbRestClient.getAllRuleChains()
                    .stream()
                    .filter(ruleChain -> ruleChain.getName().equals(RULE_CHAIN_NAME))
                    .findAny()
                    .flatMap(ruleChain -> tbRestClient.getRuleChainById(ruleChain.getUuidId()))
                    .ifPresent(ruleChain -> {
                        tbRestClient.deleteRuleChain(ruleChain.getUuidId());
                    });
```

**Comment by Deesthortered** [comment]
Anyway, the "remove" mode must make a state when the "generate" mode must work, but different rulechains with the same name make uncertainty that causes undefined behaviour. You have to remove all rulechains by given name.

**Comment by djigitiniot** [comment]
@Deesthortered should I change the logic in BasicSolution then ?

**Review by Deesthortered — APPROVED** [review]
Okay, let it be

**Pull request merged**